### PR TITLE
add `OpInfo` for `torch.nn.functional.grid_sample`

### PR DESCRIPTION
**Pull request by pmeier**
Addresses facebookresearch/functorch#78.

### CODE REVIEW

**Comment by pmeier** [comment]
We see the same failure here as observed in https://github.com/pytorch/pytorch/pull/62254#issuecomment-887561871: https://app.circleci.com/pipelines/github/pytorch/pytorch/357039/workflows/1e6f673a-d7bd-47f8-8612-0dc0ae67f249/jobs/15053582

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -7875,6 +7907,14 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
                    decorators=(toleranceOverride({torch.float32: tol(atol=0, rtol=4e-6), }),),
                    dtypes=all_types_and(torch.bool),
                    safe_casts_outputs=True),
+    OpInfo(
+        "nn.functional.grid_sample",
+        ref=_NOTHING,
+        dtypes=floating_types_and(),
```
Comment (pmeier) [inline torch/testing/_internal/common_methods_invocations.py:7913]:
```suggestion
        dtypes=floating_types(),
```

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -4547,6 +4547,38 @@ def _tensor(shape, dtype=dtype, low=None, high=None):
 
     return [SampleInput(tensor, args=args) for tensor, args in test_cases]
 
+def sample_inputs_grid_sample(op_info, device, dtype, requires_grad, **kwargs):
+    def make_tensor_(shape):
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:4551]: Nit: prefer partial

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -4547,6 +4547,38 @@ def _tensor(shape, dtype=dtype, low=None, high=None):
...
+        return make_tensor(shape, device=device, dtype=dtype, requires_grad=requires_grad)
+
+    batch_size = 2
+    num_channels = 3
+    modes = ("bilinear", "nearest")
+    align_cornerss = (False, True)
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:4557]: "cornerss" -> "corners

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -4547,6 +4547,38 @@ def _tensor(shape, dtype=dtype, low=None, high=None):
...
+    padding_modes = ("zeros", "border", "reflection")
+
+    sample_inputs = []
+    for dim in (2, 3):
+        input = make_tensor_((batch_size, num_channels, *[S] * dim))
+        grid = make_tensor_((batch_size, *[S] * dim, dim))
+
+        modes_ = (*modes, "bicubic") if dim == 2 else modes
+
+        for mode, padding_mode, align_corners in itertools.product(modes_, padding_modes, align_cornerss):
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:4567]: Nice product

File: torch/testing/_internal/common_methods_invocations.py [hunk 5]
```diff
@@ -4547,6 +4547,38 @@ def _tensor(shape, dtype=dtype, low=None, high=None):
...
+            sample_inputs.append(
+                SampleInput(
+                    input,
+                    args=(grid,),
+                    kwargs=dict(
+                        mode=mode,
+                        padding_mode=padding_mode,
+                        align_corners=align_corners,
+                    )
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:4576]: Style nit: `)))`

(but whatever)

**Comment by mruberry** [comment]
Test failures look real:

test_fn_grad_nn_functional_grid_sample_cuda_float64

```
  File "/opt/conda/lib/python3.9/site-packages/torch/autograd/gradcheck.py", line 544, in _check_analytical_jacobian_attributes
    raise GradcheckError('Backward is not reentrant, i.e., running backward with '
torch.autograd.gradcheck.GradcheckError: Backward is not reentrant, i.e., running backward with same input and grad_output multiple times gives different values, although analytical gradient matches numerical gradient.The tolerance for nondeterminism was 0.0.
```

needs to be addressed.

test_variant_consistency_jit_nn_functional_grid_sample_cuda_float32 ca be skipped, I think. cc @eellison 

```
RuntimeError: aliasOp != torch::jit::getOperatorAliasMap().end()INTERNAL ASSERT FAILED at "/var/lib/jenkins/workspace/torch/csrc/jit/passes/utils/check_alias_annotation.cpp":159, please report a bug to PyTorch. 
```

cc @zou3519 -- would you shepherd this PR through to merge?

File: torch/testing/_internal/common_methods_invocations.py (hunk 3 above)
Comment (pmeier) [reply torch/testing/_internal/common_methods_invocations.py:4557]: How are we going to differentiate it from the loop variable? Basically, we need the plural of `align_corners` here. I'm fully aware that adding an extra `s` is not grammatically correct, but it seems like an easy way out here.

Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:4557]: Oh OK. That's fine, my mistake

**Review by zou3519 — APPROVED** [review]

**Pull request closed without merging**